cassiopeia/solver/PercolationSolver.py

```diff
@@ -0,0 +1,350 @@
+                G.add_edge(i, j)
+
+        if len(G.edges) == 0:
+            return samples, []
+
+        connected_components = list(nx.connected_components(G))
+        sorted_edge_weights = sorted(edge_weight_buckets, reverse=True)
+
+        # Percolate the similarity graph by continuously removing the minimum
+        # edge until at least two components exists
+        while len(connected_components) <= 1:
+            min_weight = sorted_edge_weights.pop()
+            for edge in edge_weight_buckets[min_weight]:
+                G.remove_edge(edge[0], edge[1])
+            connected_components = list(nx.connected_components(G))
+
+        # If the number of connected components > 2, merge components by
+        # joining the most similar LCAs of each component until
+        # only 2 remain
+        partition_sides = []
+
+        if len(connected_components) > 2:
+            for c in range(len(connected_components)):
+                connected_components[c] = list(connected_components[c])
+            lcas = {}
+            component_to_nodes = {}
+            # Find the LCA of the nodes in each connected component
+            for ind in range(len(connected_components)):
+                component_identifier = "component" + str(ind)
+                component_to_nodes[component_identifier] = connected_components[ind]
+                character_vectors = [
+                    list(i)
+                    for i in list(
+                        unique_character_array[connected_components[ind], :]
+                    )
+                ]
+                lcas[component_identifier] = data_utilities.get_lca_characters(
+                    character_vectors, missing_state_indicator
+                )
+            # Build a tree on the LCA characters to cluster the components
+            lca_tree = CassiopeiaTree(
+                pd.DataFrame.from_dict(lcas, orient="index"),
+                missing_state_indicator=missing_state_indicator,
+                priors=priors,
+            )
+
+            self.joining_solver.solve(lca_tree)
+            lca_network = lca_tree.get_tree_topology()
+            grouped_components = []
+            root = [
+                n for n in lca_network if lca_network.in_degree(n) == 0
+            ][0]
+
+            # Take the split at the root as the clusters of components
+            # in the split, ignoring unifurcations
+            current_node = root
+            while len(grouped_components) == 0:
+                successors = list(lca_network.successors(current_node))
+                if len(successors) == 1:
+                    current_node = successors[0]
+                else:
+                    for i in successors:
+                        grouped_components.append(
+                            solver_utilities.get_leaf_children(
+                                lca_network, i
+                            )
+                        )
```

Comment:
instead of having this function that gets the leaves of a subtree in a network, why not use the function we already have in the `CassiopeiaTree` -- `lca_tree.leaves_in_subtree(i)`?

A similar argument can be made for using the CassiopeiaTree interface above for simple things like getting the root, but here I think it's especially worthwhile to mention that we don't need a new function in `solver_utilities` for this.